File: 2ºAÑO/1ºCuatrimestre/PDOO/examen_p1_p2/examen/pkg/Civitas_ruby-0.0.1/lib/sorpresa.rb
# encoding:utf-8

require_relative "tipo_sorpresa.rb"

module Civitas
  class Sorpresa
    
    def initialize(tipo,tablero=nil,valor=-1,texto="\0",mazo=nil)
      @tipo_sorpresa = tipo
      @texto = texto
      @valor = valor
      @mazo = mazo
      @tablero = tablero
    end
    
    def self.new_carcel(tipo, tablero)
      if tipo == Tipo_sorpresa::IRCARCEL
        new(tipo,tablero,tablero.num_casilla_carcel,"\0",nil)
      end
    end
    
    def self.new_ir_casilla(tipo, tablero, valor, texto)
      if tipo == Tipo_sorpresa::IRCASILLA
        new(tipo,tablero,valor,texto,nil)
      end
    end
    
    def self.new_generico(tipo,valor, texto)
      new(tipo,nil,valor,texto,nil)
    end
    
    def self.new_salir_carcel(tipo, mazo)
      if tipo == Tipo_sorpresa::SALIRCARCEL
        new(tipo,nil,0,"\0",mazo)
      end
    end
    
        def self.init
      @valor = -1
      @mazo = nil
      @tablero = nil
    end
    
    
    def aplicar_a_jugador(actual,todos)
      
      if @tipo_sorpresa == Tipo_sorpresa::IRCASILLA
        aplicar_a_jugador_ir_a_casilla(actual,todos)
        
        else if @tipo_sorpresa == Tipo_sorpresa::IRCARCEL
          aplicar_a_jugador_ir_carcel(actual,todos)
          
          else if @tipo_sorpresa == Tipo_sorpresa::PAGARCOBRAR
            aplicar_a_jugador_pagar_cobrar(actual,todos)
            
            else if @tipo_sorpresa == Tipo_sorpresa::PORCASAHOTEL
              aplicar_a_jugador_por_casa_hotel(actual,todos)
              
              else if @tipo_sorpresa == Tipo_sorpresa::PORJUGADOR
                aplicar_a_jugador_por_jugador(actual,todos)
                
                else if @tipo_sorpresa == Tipo_sorpresa::SALIRCARCEL
                  aplicar_a_jugador_salir_carcel(actual,todos)
                end
              end
            end
          end
        end
      end
      
    end
    
    def aplicar_a_jugador_ir_a_casilla(actual,todos)
      
      if jugador_correcto(actual,todos)
        
        informe(actual,todos)
        casilla_actual = todos[actual].num_casilla_actual
        tirada = @tablero.calcular_tirada(casilla_actual, @valor)
        posicion = @tablero.nueva_posicion(casilla_actual, tirada)
        todos[actual].mover_a_casilla(posicion)
        #@tablero.casilla.recibe_jugador(actual, todos)
        
      end
    end
    
    def aplicar_a_jugador_ir_carcel(actual,todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        todos[actual].encarcelar(@tablero.num_casilla_carcel)
      end
    end
    
    def aplicar_a_jugador_pagar_cobrar(actual,todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        todos[actual].modificar_saldo(@valor)
      end
    end
    
    def aplicar_a_jugador_por_casa_hotel(actual,todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        todos[actual].modificar_saldo(@valor*todos[actual].cantidad_casas_hoteles)
      end
    end
    
    def aplicar_a_jugador_por_jugador(actual,todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        
        s1 = Sorpresa.new_generico(Tipo_sorpresa::PAGARCOBRAR,@valor*(-1),"pagarCobrar")
        
        for i in 1..todos.size-1 do
          if i != actual 
            s1.aplicar_a_jugador_pagar_cobrar(i,todos)
          end
        end
        
        s2 = Sorpresa.new_generico(Tipo_sorpresa::PAGARCOBRAR,@valor*todos.size-1,"pagarCobrar")
        s2.aplicar_a_jugador_pagar_cobrar(actual,todos)
        
      end
    end
    
    def aplicar_a_jugador_salir_carcel(actual,todos)
      if jugador_correcto(actual,todos)
        informe(actual,todos)
        
        for i in 0..todos.size-1 do   
          if todos[i].tiene_salvoconducto
            sigue = false
          end
        end
        if sigue
          todos[actual].obtener_salvoconducto
        end
        salir_del_mazo
      end
    end
    
    def informe(actual,todos)
      Diario.instance.ocurre_evento("Se está aplicando una sorpresa al jugador " + todos[actual].nombre)
    end
    

    
    def jugador_correcto(actual,todos)
      return actual >= 0 && actual < todos.size
    end
    
    def salir_del_mazo
      if @tipo == Tipo_sorpresa::SALIRCARCEL
        @mazo.inhabilitar_carta_especial(self)  #duda de esto
      end
    end
    
    def to_s
      "Tipo sorpresa = #{@tipo_sorpresa.to_s} \n valor = #{@valor.to_s} \n texto = #{@texto.to_s}
            \n mazo = #{@mazo.to_s}  \n tablero = #{@tablero.to_s}"
    end

    def usada
      if @tipo == Tipo_sorpresa::SALIRCARCEL
        @mazo.habilitar_carta_especial(self)  #duda de esto
      end
    end
    
    protected :aplicar_a_jugador_pagar_cobrar
    
    private :informe, :aplicar_a_jugador_ir_a_casilla,
            :aplicar_a_jugador_ir_carcel,
            :aplicar_a_jugador_por_casa_hotel, :aplicar_a_jugador_por_jugador
            :aplicar_a_jugador_salir_carcel
   #publicos por defecto    
  end
end

File: 2ºAÑO/1ºCuatrimestre/PDOO/Civitas/Civitas/src/civitas/Sorpresa.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author Jose
 */
public class Sorpresa {
    private String texto;
    private int valor;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private TipoSorpresa tipo;
    
    Sorpresa(TipoSorpresa tipo, Tablero tablero){
        this.init();
        if(tipo == TipoSorpresa.IRACARCEL) {
            valor = tablero.getCarcel();
            this.tablero = tablero;
            this.tipo = tipo;
        }
    }
    
    Sorpresa(TipoSorpresa tipo, Tablero tablero, int valor, String texto){
        this.init();
        if(tipo == TipoSorpresa.IRCASILLA) {
            this.tipo = tipo;
            this.texto = texto;
            this.tablero = tablero;
            this.valor = valor;
        }
    }
        
    Sorpresa(TipoSorpresa tipo, int valor, String texto){
        this.init();
        this.tipo = tipo;
        this.valor = valor;
        this.texto = texto;
    }
    
    Sorpresa(TipoSorpresa tipo, MazoSorpresas mazo){
        this.init();
        if(tipo == TipoSorpresa.SALIRCARCEL) {
            this.tipo = tipo;
            this.mazo = mazo;
        }
    }
        
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        if(tipo == TipoSorpresa.IRCASILLA)
            aplicarAJugador_irACasilla(actual,todos);
        else if(tipo == TipoSorpresa.IRACARCEL)
            aplicarAJugador_irCarcel(actual,todos);
        else if(tipo == TipoSorpresa.PAGARCOBRAR)
            aplicarAJugador_pagarCobrar(actual,todos);
        else if(tipo == TipoSorpresa.PORCASAHOTEL)
            aplicarAJugador_porCasaHotel(actual,todos);
        else if(tipo == TipoSorpresa.PORJUGADOR)
            aplicarAJugador_porJugador(actual,todos);
        else if(tipo == TipoSorpresa.SALIRCARCEL)
            aplicarAJugador_salirCarcel(actual,todos);
    }
    
    private void aplicarAJugador_irACasilla(int actual, ArrayList<Jugador> todos){
        if(this.jugadorCorrecto(actual, todos))
        {
            this.informe(actual, todos);
            int casillaActual = todos.get(actual).getNumCasillaActual();
            int tirada = tablero.calcularTirada(casillaActual, valor);
            int posicion = tablero.nuevaPosicion(casillaActual, tirada);
            todos.get(actual).moverACasilla(posicion);
            //tablero.getCasilla(casillaActual).recibeJugador(actual, todos);
            
        }
    }
    
    private void aplicarAJugador_irCarcel(int actual, ArrayList<Jugador> todos){
        if(this.jugadorCorrecto(actual, todos))
        {
            this.informe(actual, todos);
            todos.get(actual).encarcelar(tablero.getCarcel());
        }
    }
    
    private void aplicarAJugador_pagarCobrar(int actual, ArrayList<Jugador> todos){
        if(this.jugadorCorrecto(actual, todos))
        {
            this.informe(actual, todos);
            todos.get(actual).modificarSaldo(valor);
        }
    }
    
    private void aplicarAJugador_porCasaHotel(int actual, ArrayList<Jugador> todos){
        if(this.jugadorCorrecto(actual, todos))
        {
            this.informe(actual, todos);
            todos.get(actual).modificarSaldo(valor*todos.get(actual).cantidadCasasHoteles());
        }
    }
    
    private void aplicarAJugador_porJugador(int actual, ArrayList<Jugador> todos){
        if(this.jugadorCorrecto(actual, todos))
        {
            this.informe(actual, todos);
            
            Sorpresa s1 = new Sorpresa(TipoSorpresa.PAGARCOBRAR,valor*(-1),"pagarCobrar");
            for(int i = 0; i<todos.size(); i++)
            {
                if(i != actual)
                    s1.aplicarAJugador_pagarCobrar(i, todos);
            }
            
            Sorpresa s2 = new Sorpresa(TipoSorpresa.PAGARCOBRAR,valor*todos.size()-1,"pagarCobrar");
            s2.aplicarAJugador_pagarCobrar(actual, todos);
            
        }
    }
    
    private void aplicarAJugador_salirCarcel(int actual, ArrayList<Jugador> todos){
        boolean sigue = true;
        if(this.jugadorCorrecto(actual, todos))
        {
            this.informe(actual, todos);
            for(int i = 0; i < todos.size() && sigue; i++)
            {
                if(todos.get(i).tieneSalvoConducto())
                    sigue = false;
            }
            
            if(sigue)
                todos.get(actual).obtenerSalvoconducto(this);
            this.salirDelMazo();
        }
    }
    
    private void informe(int actual, ArrayList<Jugador> todos){
        Diario.getInstance().ocurreEvento("Se está aplicando una sorpresa al jugador " + todos.get(actual).getNombre());
    }
   
    private void init(){
        valor = -1;
        mazo = null;
        tablero = null;
    }
    
    public boolean jugadorCorrecto(int actual,ArrayList<Jugador> todos ){
        return actual >= 0 && actual < todos.size() ;
    }
    
    void salirDelMazo(){
        if(tipo == TipoSorpresa.SALIRCARCEL) {
            mazo.inhabilitarCartaEspecial(this);
        }
    }

    @Override
    public String toString() {
        return "Sorpresa{" + "texto=" + texto + ", valor=" + valor + ", tablero=" + tablero + ", mazo=" + mazo + ", tipo=" + tipo + '}';
    }
    
    
    
    void usada(){
        if(tipo == TipoSorpresa.SALIRCARCEL) {
            mazo.habilitarCartaEspecial(this);
        }
    }
    
}
